concurrent/src/main/java/io/airlift/concurrent/AsyncSemaphore.java

```diff
@@ -37,26 +42,70 @@
  * to do the bulk of its work asynchronously.
  */
 @ThreadSafe
-public class AsyncSemaphore<T>
+public class AsyncSemaphore<T, R>
 {
-    private final Queue<QueuedTask<T>> queuedTasks = new ConcurrentLinkedQueue<>();
+    private final Queue<QueuedTask<T, R>> queuedTasks = new ConcurrentLinkedQueue<>();
     private final AtomicInteger counter = new AtomicInteger();
     private final Runnable runNextTask = this::runNext;
     private final int maxPermits;
     private final Executor submitExecutor;
-    private final Function<T, ListenableFuture<?>> submitter;
+    private final Function<T, ListenableFuture<R>> submitter;
+
+    /**
+     * Process a list of tasks as a single unit
+     * (similar to {@link com.google.common.util.concurrent.Futures#allAsList(ListenableFuture[])})
+     * with limiting the number of tasks running in parallel.
+     * <p>
+     * This method may be useful for limiting the number of concurrent requests sent to a remote server when
+     * trying to load multiple related entities concurrently:
+     * <p>
+     * For example:
+     * <pre>{@code
+     * List<Integer> userIds = Lists.of(1, 2, 3);
+     * ListenableFuture<List<UserInfo>> future = processAll(ids, client::getUserInfoById, 2, executor);
+     * List<UserInfo> userInfos = future.get(...);
+     * }</pre>
+     *
+     * @param tasks tasks to process
+     * @param submitter task submitter
+     * @param maxPermits maximum number of tasks permitted to run in parallel
+     * @param executor task submission executor
+     * @return {@link ListenableFuture} containing a list of values returned by the {@code tasks}.
+     * The order of elements in the list matches the order of {@code tasks}.
+     * If the result future is cancelled all the remaining tasks are cancelled (submitted tasks will be cancelled, pending tasks will not be submitted).
+     * If any of the submitted tasks fails or is canceled, the result future is too.
+     * If any of the submitted tasks fails or is canceled, the remaining tasks are cancelled.
+     */
+    public static <T, R> ListenableFuture<List<R>> processAll(List<T> tasks, Function<T, ListenableFuture<R>> submitter, int maxPermits, Executor executor)
+    {
+        AtomicReference<ListenableFuture<List<R>>> resultFutureReference = new AtomicReference<>();
+        AsyncSemaphore<T, R> semaphore = new AsyncSemaphore<>(maxPermits, executor, task -> {
+            ListenableFuture<List<R>> resultFuture = resultFutureReference.get();
+            if (resultFuture != null && resultFuture.isDone()) {
+                return immediateCancelledFuture();
+            }
```

Comment:
Since you are already doing `if (!queuedTask.getCompletionFuture().isDone()) {` in `runNext()`, you actually don't need to do this here with resultFutureReference. If allAsList is explicitly cancelled, or if any of the sub futures fail, everything that hasn't run yet will be cancelled (and will thus be done). There will be no additional value to return a `immediateCancelledFuture()` here.

Reply:
This is to cover a corner case I discovered when testing. When the executor provided is a `directExecutor()` cancelling the result future will attempt to synchronously cancel underlying futures triggering the `releasePermit` callback on every future resulting in unnecessary submissions.

Reply:
@arhimondr, ok yea, I don't think there is a good way around this other than what you suggested. But I would suggest this minor refactoring of this function:

```java
    public static <T, R> ListenableFuture<List<R>> processAll(List<T> tasks, Function<T, ListenableFuture<R>> submitter, int maxConcurrency, Executor submitExecutor)
    {
        SettableFuture<List<R>> resultFuture = SettableFuture.create();
        AsyncSemaphore<T, R> semaphore = new AsyncSemaphore<>(maxConcurrency, submitExecutor, task -> {
            if (resultFuture.isCancelled()) {
                // Task cancellation tends to happen in task submission order, which can race with subsequent task submissions after previous cancellations.
                // This check prevents this race ever from occurring.
                return immediateCancelledFuture();
            }
            return submitter.apply(task);
        });
        resultFuture.setFuture(allAsListWithFailureCancellation(tasks.stream()
                .map(semaphore::submit)
                .collect(toImmutableList())));
        return resultFuture;
    }

    /**
     * Same as {@link com.google.common.util.concurrent.Futures#allAsList(ListenableFuture[])}, but additionally cancels any remaining component Futures
     * if the result has already failed.
     *
     * @param futures futures to combine
     * @return a future that provides a list of the results of the component futures
     */
    private static <V> ListenableFuture<List<V>> allAsListWithFailureCancellation(Iterable<? extends ListenableFuture<? extends V>> futures)
    {
        List<ListenableFuture<? extends V>> futuresSnapshot = ImmutableList.copyOf(futures);
        ListenableFuture<List<V>> listFuture = allAsList(futuresSnapshot);
        addExceptionCallback(listFuture, () -> futuresSnapshot.forEach(future -> future.cancel(true)));
        return listFuture;
    }
```

I would also greatly appreciate if you put allAsListWithFailureCancellation into MoreFutures, since this is a behavior that I would very much like to reuse in the future.